File: RecyclerViewProduct/app/src/main/java/com/example/recyclerviewproduct/ProductsAdapter.java
package com.example.recyclerviewproduct;

import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

public class ProductsAdapter extends RecyclerView.Adapter<ProductsAdapter.ProductViewHolder> {

    ImageView productImageView;
    TextView productTitleTxtView, productPriceTxtView;
    ArrayList<Product> products;

    public ProductsAdapter(ArrayList<Product> products1){
        this.products = products1;
    }

    class ProductViewHolder extends RecyclerView.ViewHolder{

        public ProductViewHolder(@NonNull View itemView) {
            super(itemView);

            productImageView = itemView.findViewById(R.id.productImageView);
            productTitleTxtView = itemView.findViewById(R.id.productTitleTxtView);
            productPriceTxtView = itemView.findViewById(R.id.productPriceTxtView);

            productImageView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    //way 3
                    Intent intent = new Intent(v.getContext(), ProductDetailsActivity.class);
                    intent.putExtra("product",products.get(getAdapterPosition()));
                    v.getContext().startActivity(intent);
                }
            });
        }
    }

    @NonNull
    @Override
    public ProductViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        View productView = layoutInflater.inflate(R.layout.product_view_holder,null);

        productImageView = productView.findViewById(R.id.productImageView);
        productTitleTxtView = productView.findViewById(R.id.productTitleTxtView);
        productPriceTxtView = productView.findViewById(R.id.productPriceTxtView);

//        productImageView.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                Toast.makeText(parent.getContext(),"Image View Clicked" , Toast.LENGTH_LONG).show();
//            }
//        });
//
//        productTitleTxtView.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                Toast.makeText(parent.getContext(),"Title Text View Clicked", Toast.LENGTH_LONG).show();
//            }
//        });
//
//        productPriceTxtView.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                Toast.makeText(parent.getContext(),"Price Text View Clicked", Toast.LENGTH_LONG).show();
//            }
//        });

        return new ProductViewHolder(productView);
    }

    @Override
    public void onBindViewHolder(@NonNull ProductViewHolder holder, int position) {
        Product product = products.get(position);       //extract eachProduct from array of product objects

        //get views from holder bu using findViewById
        productImageView = holder.itemView.findViewById(R.id.productImageView);
        productTitleTxtView = holder.itemView.findViewById(R.id.productTitleTxtView);
        productPriceTxtView = holder.itemView.findViewById(R.id.productPriceTxtView);

        //bind eachProduct data with the corresponding views
        productTitleTxtView.setText(product.productTitle);
        productPriceTxtView.setText(product.productPrice + " ");
        productImageView.setImageResource(R.drawable.ic_launcher_background);

        //way 2 - attaching listeners inside onBindViewHolder method
        /*
        productImageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                    Toast.makeText(v.getContext(),"ImageViewClicked  " + product.productId, Toast.LENGTH_SHORT).show();
            }
        });

        productTitleTxtView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Toast.makeText(v.getContext(),"TitleTextViewClicked  " + product.productTitle, Toast.LENGTH_SHORT).show();

                Intent intent = new Intent(v.getContext(), ProductDetailsActivity.class);
                intent.putExtra("product_title",product.productTitle);
                intent.putExtra("product_price", product.productPrice);
                intent.putExtra("product_image", product.imageId);
                intent.putExtra("product_id", product.productId);

                v.getContext().startActivity(intent);
            }
        });

        productPriceTxtView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                    Toast.makeText(v.getContext(),"PriceTextViewClicked  " + product.productPrice, Toast.LENGTH_SHORT).show();
            }
        }); */
    }

    @Override
    public int getItemCount() {
        return products.size();
    }
}

File: RecyclerViewKotlin/app/src/main/java/com/example/recyclerviewkotlin/ProductsAdapter.kt
package com.example.recyclerviewkotlin

import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder

class ProductsAdapter(private var products : ArrayList<Product>, private var users : ArrayList<User>) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    val PRODUCT_VIEW_TYPE = 1
    val USER_VIEW_TYPE = 2

    inner class ProductViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private lateinit var txtViewProductTitle : TextView
        private lateinit var txtViewProductPrice : TextView
        private lateinit var imgView1 : ImageView

        init {
            txtViewProductTitle = itemView.findViewById(R.id.txtViewProductTitle)
            txtViewProductPrice = itemView.findViewById(R.id.txtViewProductPrice)
            imgView1 = itemView.findViewById(R.id.imgView1)

            txtViewProductTitle.setOnClickListener(View.OnClickListener {
                var intent : Intent = Intent(it.context,ProductDetailsActivity::class.java)
                val eachProduct = products[adapterPosition]
                intent.putExtra("product",eachProduct)
                itemView.context.startActivity(intent)
            })
        }
    }

    inner class UserViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private lateinit var txtViewUsername : TextView
        init {
            txtViewUsername = itemView.findViewById(R.id.txtViewUsername)
        }
    }

    override fun getItemViewType(position: Int): Int {
        if (position % 2 == 0){
            return PRODUCT_VIEW_TYPE
        }
        return USER_VIEW_TYPE
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        if (viewType == PRODUCT_VIEW_TYPE){
            val layoutInflater = LayoutInflater.from(parent.context)
            val productView = layoutInflater.inflate(R.layout.product_view_holder,null)

            return ProductViewHolder(productView)
        }

        val layoutInflater = LayoutInflater.from(parent.context)
        val userView = layoutInflater.inflate(R.layout.user_view_holder,null)

        return UserViewHolder(userView)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is ProductViewHolder){
            var product = products[position / 2]
            var txtViewProductTitle = holder.itemView.findViewById<TextView>(R.id.txtViewProductTitle)
            txtViewProductTitle.text = product.productTitle

            holder.itemView.findViewById<TextView>(R.id.txtViewProductPrice).text = "" + product.productPrice
            holder.itemView.findViewById<ImageView>(R.id.imgView1).setImageResource(R.drawable.ic_launcher_background)
        } else {
            var user = users[(position - 1) / 2]
            var txtViewUsername = holder.itemView.findViewById<TextView>(R.id.txtViewUsername)

            txtViewUsername.text = user.userName
        }
    }

    override fun getItemCount(): Int {
        return products.size + users.size
    }
}